### PR TITLE
feat: add profiling route handler to debug runtime

### PR DESCRIPTION
**Pull request by alexstan12**
**What type of PR is this?**
feature

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rchincha** [comment]
This is actually useful. How much runtime overhead does it add? 
https://medium.com/google-cloud/continuous-profiling-of-go-programs-96d4416af77b

**Comment by rchincha** [comment]
This is actually useful. How much runtime overhead does it add? 
https://medium.com/google-cloud/continuous-profiling-of-go-programs-96d4416af77b

**Comment by alexstan12** [comment]
> This is actually useful. How much runtime overhead does it add? https://medium.com/google-cloud/continuous-profiling-of-go-programs-96d4416af77b

so far zb has not provided any relevant data regarding dfif in performance between zot with debug build tag (that include profiling) and zot w/o debug. Depending on the test run, sometimes they seem to have almost the same performance, sometimes they have half the number of requests per second.

More tests are to be done in order to establish a clear result.

**Comment by alexstan12** [comment]
> This is actually useful. How much runtime overhead does it add? https://medium.com/google-cloud/continuous-profiling-of-go-programs-96d4416af77b

so far zb has not provided any relevant data regarding dfif in performance between zot with debug build tag (that include profiling) and zot w/o debug. Depending on the test run, sometimes they seem to have almost the same performance, sometimes they have half the number of requests per second.

More tests are to be done in order to establish a clear result.

File: pkg/debug/pprof/pprof.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	"zotregistry.io/zot/pkg/api/config"
+	"zotregistry.io/zot/pkg/log" //nolint:goimports
+)
+
+func SetupPprofRoutes(conf *config.Config, router *mux.Router, authFunc mux.MiddlewareFunc,
+	log log.Logger,
+) {
+	// router has authn mdw already applied
+	pprofRouter := router.PathPrefix("/_zot/debug/pprof").Subrouter()
+	pprofRouter.Methods(http.MethodGet).Handler(http.HandlerFunc(
```
Comment (andaaron) [inline pkg/debug/pprof/pprof.go:21]: Let's check if the `trace` and `goroutines` profiles are useful in our context as well.

File: pkg/debug/pprof/pprof.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+	"zotregistry.io/zot/pkg/api/config"
+	"zotregistry.io/zot/pkg/log" //nolint:goimports
+)
+
+func SetupPprofRoutes(conf *config.Config, router *mux.Router, authFunc mux.MiddlewareFunc,
+	log log.Logger,
+) {
+	// router has authn mdw already applied
+	pprofRouter := router.PathPrefix("/_zot/debug/pprof").Subrouter()
```
[the remark this answers is not in the document]
Comment (alexstan12) [reply pkg/debug/pprof/pprof.go:20]: done, replaced with constant.

File: pkg/debug/pprof/pprof.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# CPU Profiling in Zot
+
+This project gives the user the posibility to debug and profile the runtime to find CPU intensive function calls. An in-depth article on profiling in Go can be found [here](https://go.dev/blog/pprof).
+This endpoint `/_zot/debug/pprof` can be queried to gather information of the runtime. 
+
+For example, the following can be used to gather the cpu profile for the amount of seconds specified as a query parameter, and then the results are stored in `cpu.prof` file:
+```
+curl -sK -v http://localhost:8080/v2/_zot/debug/pprof?seconds=30 > cpu.prof
```
[the remark this answers is not in the document]
Comment (alexstan12) [reply pkg/debug/pprof/pprof.md]: updated

**Review by andaaron — APPROVED** [review]

File: pkg/debug/constants/consts.go [hunk 4]
```diff
@@ -3,4 +3,5 @@ package constants
 const (
 	Debug                 = "/_zot/debug"
 	GQLPlaygroundEndpoint = Debug + "/graphql-playground"
+	ProfilingEndpoint     = "/_zot/pprof/"
```
Comment (rchincha) [inline pkg/debug/constants/consts.go:6]: Endpoints need to be sorted out.

File: pkg/api/routes.go [hunk 5]
```diff
@@ -118,6 +119,9 @@ func (rh *RouteHandler) SetupRoutes() {
 	// swagger
 	debug.SetupSwaggerRoutes(rh.c.Config, rh.c.Router, AuthHandler(rh.c), rh.c.Log)
 
+	// pprof
+	pprof.SetupPprofRoutes(rh.c.Config, prefixedRouter, AuthHandler(rh.c), rh.c.Log)
```
Comment (rchincha) [inline pkg/api/routes.go:123]: Only "admin" persona should be able do this.

File: pkg/debug/pprof/pprof.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+	"cmdline":      "The command line invocation of the current program",
+	"goroutine":    "Stack traces of all current goroutines. Use debug=2 as a query parameter to export in the same format as an unrecovered panic.",  //nolint: lll
+	"heap":         "A sampling of memory allocations of live objects. You can specify the gc GET parameter to run GC before taking the heap sample.", //nolint: lll
+	"mutex":        "Stack traces of holders of contended mutexes",
+	"profile":      "CPU profile. You can specify the duration in the seconds GET parameter. After you get the profile file, use the go tool pprof command to investigate the profile.", //nolint: lll
+	"threadcreate": "Stack traces that led to the creation of new OS threads",
+	"trace":        "A trace of execution of the current program. You can specify the duration in the seconds GET parameter. After you get the trace file, use the go tool trace command to investigate the trace.", //nolint: lll
+}
+
+func SetupPprofRoutes(conf *config.Config, router *mux.Router, authFunc mux.MiddlewareFunc,
```
Comment (andaaron) [inline pkg/debug/pprof/pprof.go:45]: We need to check if the user is an admin before running this.

File: pkg/debug/pprof/pprof.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
+//go:build debug
+// +build debug
```
Comment (andaaron) [inline pkg/debug/pprof/pprof.go:2]: Use a different build tag for pprof

**Comment by rchincha** [comment]
Isolate this under a "-profile" build tag for perf numbers.

**Comment by rchincha** [comment]
https://man7.org/linux/man-pages/man1/perf-top.1.html

^ maybe a better way to introspect.

**Comment by rchincha** [comment]
Attach the PNG to this issue and remove from the commit.

File: pkg/debug/pprof/flamegraph.png
Comment (andaaron) [inline pkg/debug/pprof/flamegraph.png:1]: Put this in the issue itself, not the committed code.

**Comment by alexstan12** [comment]
> https://man7.org/linux/man-pages/man1/perf-top.1.html
> 
> ^ maybe a better way to introspect.

I ran the perf stat tool with a timeout of 1 min on two separate zot servers, one built with the 'binary' make target and one that adds the profiling endpoints (and has the profile build tag).
`timeout -s SIGINT 1m perf stat -d ./bin/zot-linux-amd64 serve ./examples/config-example.json`
`timeout -s SIGINT 1m perf stat -d ./bin/zot-linux-amd64-profile serve ./examples/config-example.json`
The results are similar, indicating **no passive overhead** in the binary that includes profiling - i.e. when no profiling endpoints are called.

Below you can see the results, with emphasis on the most relevant values.
perf

**Comment by alexstan12** [comment]
flamegraph

**Comment by andaaron** [comment]
Superseded by https://github.com/project-zot/zot/pull/1818/.

**Pull request closed without merging**